File: db4o.net/Db4objects.Db4o.Tests/Db4objects.Db4o.Tests/Common/Constraints/UniqueFieldValueDoesNotThrowTestCase.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using System.Collections;
using Db4oUnit.Extensions;
using Db4oUnit.Extensions.Fixtures;
using Db4objects.Db4o.Config;
using Db4objects.Db4o.Constraints;
using Db4objects.Db4o.Tests.Common.Constraints;

namespace Db4objects.Db4o.Tests.Common.Constraints
{
	public class UniqueFieldValueDoesNotThrowTestCase : AbstractDb4oTestCase, ICustomClientServerConfiguration
	{
		public class Item
		{
			public long id;

			public string name;

			public Item()
			{
			}

			public Item(int id, string name)
			{
				this.id = System.Convert.ToInt64(id);
				this.name = name;
			}

			public override int GetHashCode()
			{
				return id.GetHashCode();
			}
		}

		public class Holder
		{
			public Hashtable _items = new Hashtable();

			public virtual void Add(UniqueFieldValueDoesNotThrowTestCase.Item item)
			{
				_items[item] = item.id;
			}
		}

		public virtual void ConfigureClient(IConfiguration config)
		{
		}

		/// <exception cref="System.Exception"></exception>
		public virtual void ConfigureServer(IConfiguration config)
		{
			Configure(config);
		}

		/// <exception cref="System.Exception"></exception>
		protected override void Configure(IConfiguration config)
		{
			config.ObjectClass(typeof(UniqueFieldValueDoesNotThrowTestCase.Item)).ObjectField
				("id").Indexed(true);
			config.Add(new UniqueFieldValueConstraint(typeof(UniqueFieldValueDoesNotThrowTestCase.Item
				), "id"));
			config.ObjectClass(typeof(UniqueFieldValueDoesNotThrowTestCase.Holder)).CallConstructor
				(true);
		}

		/// <exception cref="System.Exception"></exception>
		public virtual void Test()
		{
			Store(NewHolder(new string[] { "foo", "bar" }));
			Db().Commit();
		}

		private object NewHolder(string[] names)
		{
			UniqueFieldValueDoesNotThrowTestCase.Holder holder = new UniqueFieldValueDoesNotThrowTestCase.Holder
				();
			for (int i = 0; i < names.Length; i++)
			{
				holder.Add(new UniqueFieldValueDoesNotThrowTestCase.Item(i, names[i]));
			}
			return holder;
		}
	}
}

File: db4oj.tests/src/com/db4o/db4ounit/common/constraints/UniqueFieldValueDoesNotThrowTestCase.java
/* Copyright (C) 2009  Versant Inc.   http://www.db4o.com */
package com.db4o.db4ounit.common.constraints;

import java.util.*;

import com.db4o.config.*;
import com.db4o.constraints.*;

import db4ounit.extensions.*;
import db4ounit.extensions.fixtures.*;

public class UniqueFieldValueDoesNotThrowTestCase
	extends AbstractDb4oTestCase
	implements CustomClientServerConfiguration {
	
	public static class Item {
		public Long id;
		public String name;
		
		public Item() {
		}
		
		public Item(int id, String name) {
			this.id = new Long(id);
			this.name = name;
		}
		
		@Override
		public int hashCode() {
			return id.hashCode();
		}
	}
	
	public static class Holder {
		public HashMap<Item, Long> _items = new HashMap<Item, Long>();
		
		public void add(Item item) {
			_items.put(item, item.id);
		}
	}
	
	public void configureClient(Configuration config) {
    }

	public void configureServer(Configuration config) throws Exception {
		configure(config);
    }	
	
	@Override
	protected void configure(Configuration config) throws Exception {
		config.objectClass(Item.class).objectField("id").indexed(true);
		
		config.add(new UniqueFieldValueConstraint(Item.class, "id"));
		config.objectClass(Holder.class).callConstructor(true);
	}
	
	public void test() throws Exception {
		store(newHolder("foo", "bar"));
		db().commit();
	}


	private Object newHolder(String... names) {
		final Holder holder = new Holder();
		for(int i = 0; i < names.length; i++) {
			holder.add(new Item(i, names[i]));
		}
		
		return holder;
	}
}
